web/modules/custom/joinup_core/src/EventSubscriber/RefreshCachedFieldsEventSubscriber.php

```diff
@@ -241,38 +210,119 @@ protected function getMatomoMethod($bundle) {
    * @return string
    *   The URL parameter name. Defaults to 'pageUrl'.
    */
-  protected function getUrlParameterName($bundle) {
+  protected function getUrlParameterName($bundle): string {
     $method = $this->getMethod($bundle);
     return $method === 'download_counts' ? 'downloadUrl' : 'pageUrl';
   }
 
   /**
    * Returns the absolute URL to the canonical page of the given entity.
    *
-   * @param \Drupal\Core\Entity\ContentEntityInterface $entity
+   * @param \Drupal\Core\Entity\EntityInterface $entity
    *   The entity for which to return the URL.
    *
    * @return string
-   *   The URL.
+   *   The entity URL.
    */
-  protected function getEntityUrl(ContentEntityInterface $entity) {
+  protected function getEntityUrl(EntityInterface $entity): string {
     return $entity->toUrl()->setAbsolute()->toString();
   }
 
+  /**
+   * Generates an argument array that represents the time range.
+   *
+   * @param int $period
+   *   The starting period represented by the number of days ago.
+   *
+   * @return array
+   *   An array with the starting and ending time range.
+   */
+  protected function getDateRange($period): array {
+    return [
+      // If the period is 0 we should get all results since launch.
+      $period > 0 ? (new DateTimePlus("$period days ago"))->format('Y-m-d') : $this->configFactory->get('joinup_core.matomo_settings')->get('launch_date'),
+      (new DateTimePlus())->format('Y-m-d'),
+    ];
+  }
+
+  /**
+   * Filters out invalid entries from a collection of items.
+   *
+   * Invalid entries can be items that their entities don't exist or have not
+   * expired yet.
+   *
+   * @param \Drupal\cached_computed_field\ExpiredItemCollection $items
+   *   A collection of items.
+   *
+   * @return \Drupal\cached_computed_field\ExpiredItemCollection
+   *   The filtered collection of items.
+   */
+  protected function filterInvalidItems(ExpiredItemCollection $items): ExpiredItemCollection {
+    $valid_items = [];
+    foreach ($items as $item) {
+      if (!$entity = $this->getEntity($item)) {
+        continue;
+      }
+
+      // Only refresh the field if it has actually expired. It might have been
+      // updated already since it has been added to the processing queue.
+      if (!$this->fieldNeedsRefresh($item)) {
+        continue;
+      }
+
+      // Catch the case where the URL is not available. This can currently
+      // happen only when an entity is malformed, for which we already check,
+      // or the distribution we are attempting to track does not have a file.
+      if (empty($this->getUrlParameter($entity))) {
+        $this->loggerFactory->get('joinup_core')->error('Invalid url parameter: ' . $this->getUrlParameter($entity));
```

Comment:
Fixed it for the specific case. We will instead retrieve the bundle and the ID. We can assume the rest easily. The distributions are for download URLs, and nodes are for visit counts. But as the description says, this might only occur for distributions.